File: src/iterators/vector/particles/java/Pair.java
/*
 * Algorithms and Complexity                            December 01, 2022
 * IST 4310
 * Prof. M. Diaz-Maldonado
 *
 *
 * Synopsis:
 * Defines the Pair (of Points) Class.
 *
 *
 * Copyright (c) 2022 Misael Diaz-Maldonado
 * This file is released under the GNU General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 *
 * References:
 * [0] A Koenig and B Moo, Accelerated C++ Practical Programming by Example
 *
 */

public final class Pair implements Comparable<Pair>
{

  // component(s):


  private final Point first;			// (smaller) first point
  private final Point second;			// second point
  private final double distance;		// separating distance


  // constructor(s):


  Pair ()					// default constructor
  {
    double inf = Double.POSITIVE_INFINITY;
    this.first = new Point(0, 0);
    this.second = new Point(inf, 0);
    this.distance = inf;
  }


  // Synopsis: constructs from a pair of points and their distance
  Pair (Point P, Point Q, double distance)
  {
    // places the smaller Point first by design
    if (P.compareTo(Q) < 0)
    {
      this.first = P;
      this.second = Q;
      this.distance = distance;
    }
    else
    {
      this.first = Q;
      this.second = P;
      this.distance = distance;
    }
  }


  Pair (Pair pair)				// copy constructor
  {
    this.first = new Point(pair.first);
    this.second = new Point(pair.second);
    this.distance = pair.distance;
  }


  // getters:


  public Point getFirst ()			// returns the first point
  {
    return this.first;
  }


  public Point getSecond ()			// returns the second point
  {
    return this.second;
  }


  public double getDistance ()			// returns the separating distance
  {
    return this.distance;
  }


  // methods:


  //  int compareTo (Pair pair)
  //
  //  Synopsis:
  //  Implements the comparable interface.
  //  Compares Pair objects by their separating distance.
  //
  //  Input:
  //  pair	a pair object (referred to as that pair)
  //
  //  Output:
  //  comp	0 if `this pair' is equal to `that pair',
  //   		1 if `this pair' is greater than `that pair',
  //   		and -1 if `this pair' is less than `that pair'
  //
  //  COMMENTS:
  //  Because we assign integral values to the coordinates of the points
  //  and these have exact binary floating-point representations, we can
  //  compare doubles as done here with confidence.
  //


  @Override
  public int compareTo (Pair pair)
  {
    // computes the difference in their distances
    double diff = (this.distance - pair.distance);

    if (diff == 0.0)
    {
      return 0;
    }
    else if (diff < 0.0)
    {
      return -1;
    }
    else
    {
      return 1;
    }
  }


  //  boolean equalTo (Pair pair)
  //
  //  Synopsis:
  //  Defines what it means for two Pair objects to be equal to one another.
  //  Two pair objects are equal if both the first and second points that comprise them
  //  are equal.
  //
  //  Input:
  //  pair		a pair object (referred to as `that pair')
  //
  //  Output:
  //  equal		true if `this pair' and `that pair' are equal,
  //   			false otherwise
  //
  //  COMMENTS:
  //  Since we make sure that the first point is smaller than the second
  //  point when constructing a Pair we need not to consider other cases.


  public boolean equalTo (Pair pair)
  {
    // checks if the first points are equal
    boolean first = this.first.compareTo(pair.first) == 0;
    // checks if the second points are equal
    boolean second = this.second.compareTo(pair.second) == 0;

    if (first && second)
      return true;
    else
      return false;
  }


  //  Pair min (Pair first, Pair second)
  //
  //  Synopsis:
  //  Returns the smallest of the given Pairs.
  //
  //  Inputs:
  //  first		first pair
  //  second 		second pair
  //
  //  Output:
  //  min		the smallest of the two pairs


  public static Pair min (Pair first, Pair second)
  {
    if (first.compareTo(second) < 0)
      return first;
    else
      return second;
  }
}

File: src/algorithms/divide-and-conquer/closest-pair/clang/oop-closest-pair/clang/pair.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>
#include "pair.h"


static size_t get_first (const void* vclosestPair)
{
  const pair_t* closestPair = vclosestPair;
  return closestPair -> _first;
}


static size_t get_second (const void* vclosestPair)
{
  const pair_t* closestPair = vclosestPair;
  return closestPair -> _second;
}


static double get_distance (const void* vclosestPair)
{
  const pair_t* closestPair = vclosestPair;
  return closestPair -> _dist;
}


static void setter (void* vclosestPair,
		    size_t const first,
		    size_t const second,
		    double const dist)
{
  pair_t* closestPair = vclosestPair;
  if (first < second)
  {
    closestPair -> _first = first;
    closestPair -> _second = second;
    closestPair -> _dist = dist;
  }
  else
  {
    closestPair -> _first = second;
    closestPair -> _second = first;
    closestPair -> _dist = dist;
  }
}


static void minimum (void* vclosestPair,
		     const void* vclosestPairLeft,
		     const void* vclosestPairRight)
{
  pair_t* closestPair = vclosestPair;
  const pair_t* closestPairLeft = vclosestPairLeft;
  const pair_t* closestPairRight = vclosestPairRight;

  double dmin = INFINITY;
  size_t first = 0xffffffffffffffff;
  size_t second = 0xffffffffffffffff;
  double const minDistLeft = closestPairLeft -> _dist;
  double const minDistRight = closestPairRight -> _dist;
  if (minDistLeft < minDistRight)
  {
    dmin = minDistLeft;
    first = closestPairLeft -> _first;
    second = closestPairLeft -> _second;
  }
  else
  {
    dmin = minDistRight;
    first = closestPairRight -> _first;
    second = closestPairRight -> _second;
  }
  closestPair -> set(closestPair, first, second, dmin);
}


static void logger (const void* vclosestPair)
{
  const pair_t* closestPair = vclosestPair;
  size_t const first = closestPair -> _first;
  size_t const second = closestPair -> _second;
  double const distance = closestPair -> _dist;
  printf("first: %lu second: %lu distance: %e\n", first, second, distance);
}


static bool comparator (const void* vclosestPair1, const void* vclosestPair2)
{
  const pair_t* closestPair1 = vclosestPair1;
  const pair_t* closestPair2 = vclosestPair2;

  size_t const i = closestPair1 -> _first;
  size_t const j = closestPair1 -> _second;
  double const d1 = closestPair1 -> _dist;

  size_t const n = closestPair2 -> _first;
  size_t const m = closestPair2 -> _second;
  double const d2 = closestPair2 -> _dist;

  return ( ( (i == n) && (j == m) && (d1 == d2) )? true : false );
}


static pair_t* construct ()
{
  pair_t* closestPair = malloc( sizeof(pair_t) );
  if (closestPair == NULL)
  {
    printf("construct(): insufficient memory to allocate the closest pair\n");
    return NULL;
  }

  closestPair -> _first = 0xffffffffffffffff;
  closestPair -> _second = 0xffffffffffffffff;
  closestPair -> _dist = INFINITY;
  closestPair -> set = setter;
  closestPair -> min = minimum;
  closestPair -> cmp = comparator;
  closestPair -> log = logger;
  closestPair -> getFirst = get_first;
  closestPair -> getSecond = get_second;
  closestPair -> getDistance = get_distance;
  return closestPair;
}


static pair_t* deconstruct (pair_t* closestPair)
{
  if (closestPair == NULL)
  {
    return closestPair;
  }

  free(closestPair);
  closestPair = NULL;
  return closestPair;
}


pair_namespace_t const pair = {
  .create = construct,
  .destroy = deconstruct
};


/*

Algorithms and Complexity					August 24, 2023

source: pair.c
author: @misael-diaz

Synopsis:
Defines methods for the pair type.

Copyright (c) 2023 Misael Diaz-Maldonado
This file is released under the GNU General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

References:
[0] A Koenig and B Moo, Accelerated C++ Practical Programming by Example.
[1] JJ McConnell, Analysis of Algorithms, second edition.

*/
